### ISSUE TITLE
Premature timeout error during truncate

### ISSUE THREAD

**Issue by ShlomiBalalis** [scylladb/scylla#4924]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash):3.0.10-0.20190815.b3bfd8c08
Cluster size:4
OS (RHEL/CentOS/Ubuntu/AWS AMI):ami-0feaef6667612ad07

When the following truncate command `cqlsh --no-color  --request-timeout=120 --connect-
timeout=5 -e "TRUNCATE ks_truncate.standard1" 10.0.163.10 9042` was executed in my cluster, it failed with the following stderr:
```
ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)
```
The weird thing is that the command timed out only after 13 seconds, when the expected timeout was 120 seconds. Also, I could not find any reference to the command in the node's log (scratch.scylladb.com/shlomib/longevity-200gb-48h-verify-limited-node-3-log.txt)

**Comment by fgelcer** [comment]
had the similar issue, where in my case it timed out after ~100 seconds.
the node's log can be found [here](http://scratch.scylladb.com/fabio/db_node_1)

**Comment by slivne** [comment]
We need a system this reproduces on to try and add debug information to the runs ....

**Comment by glommer** [comment]
I fixed this in master more than 6 months ago.

Not backported to 3.0 due to the usual (not enough open source exposure)

**Comment by slivne** [comment]
3.0 doesn't suffer from this

3.1 does - can you refer to a commit / issue

**Comment by slivne** [comment]
Is it this 889f31fabebe0e03277eddf0668ad47195870e75

**Comment by glommer** [comment]
yes sir

**Comment by slivne** [comment]
@Avi Kivity <avi@scylladb.com> objections on backporting this to 3.1 at
this point ?

**Comment by yarongilor** [comment]
reproduced in 50gb rc4 as well:
```
< t:2019-09-05 17:08:43,644 f:nemesis.py      l:1082 c:sdcm.nemesis         p:ERROR > Command: 'cqlsh --no-color -u cassandra -p cassandra  --request-timeout=120 --connect-timeout=5 -e "TRUNCATE ks_truncate.standard1" 10.0.76.1 9042'
< t:2019-09-05 17:08:43,644 f:nemesis.py      l:1082 c:sdcm.nemesis         p:ERROR > <stdin>:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)
< t:2019-09-05 17:08:43,648 f:nemesis.py      l:1094 c:sdcm.nemesis         p:DEBUG > sdcm.nemesis.ChaosMonkey: TruncateMonkey Node longevity-tls-50gb-4d-3-1-db-node-729a1e65-4 [52.213.184.109 | 10.0.76.1] (seed: False) duration -> 99 s
```

**Comment by slivne** [comment]
@avikivity ping

objections on merging 889f31f at this point - we can do this post 3.1.0

**Comment by yarongilor** [comment]
Reproduced in version: 3.1.0.rc9-0.20191006.05272c53e
AMI: ami-09c8d17ff81a2773e
log error:
```
(DisruptionEvent Severity.ERROR): type=end name=TruncateMonkey node=Node longevity-tls-50gb-4d-3-1-db-node-5467dbcb-3 [34.240.156.240 | 10.0.104.24] (seed: False) duration=115 error=Encountered a bad command exit code!

Command: 'cqlsh --no-color -u cassandra -p cassandra  --request-timeout=120 --connect-timeout=5 -e "TRUNCATE ks_truncate.standard1" 10.0.104.24 9042'

Exit code: 2

Stdout:



Stderr:

:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)


Traceback (most recent call last):
  File "/sct/sdcm/nemesis.py", line 1067, in wrapper
    result = method(*args, **kwargs)
  File "/sct/sdcm/nemesis.py", line 1276, in disrupt
    self.call_random_disrupt_method()
  File "/sct/sdcm/nemesis.py", line 588, in call_random_disrupt_method
    disrupt_method()
  File "/sct/sdcm/nemesis.py", line 653, in disrupt_truncate
    self.target_node.run_cqlsh(cmd='TRUNCATE {}.{}'.format(keyspace_truncate, table), timeout=120)
  File "/sct/sdcm/cluster.py", line 2151, in run_cqlsh
    verbose=verbose)
  File "/sct/sdcm/remote.py", line 226, in run
    result = _run()
  File "/sct/sdcm/utils/common.py", line 87, in inner
    return func(*args, **kwargs)
  File "/sct/sdcm/remote.py", line 217, in _run
    watchers=watchers, timeout=timeout)
  File "", line 2, in run
  File "/usr/lib/python2.7/site-packages/fabric/connection.py", line 30, in opens
    return method(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/fabric/connection.py", line 721, in run
    return self._run(self._remote_runner(), command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/context.py", line 101, in _run
    return runner.run(command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/runners.py", line 291, in run
    return self._run_body(command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/runners.py", line 442, in _run_body
    raise UnexpectedExit(result)
UnexpectedExit: Encountered a bad command exit code!

Command: 'cqlsh --no-color -u cassandra -p cassandra  --request-timeout=120 --connect-timeout=5 -e "TRUNCATE ks_truncate.standard1" 10.0.104.24 9042'

Exit code: 2

Stdout:



Stderr:

:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)


```
log file: scratch.scylladb.com/qa_issues/sct_11_37_07_10_2019.log.gz

**Comment by roydahan** [comment]
We have this issue also in master, so it may be a different one than what @glommer fixed or the fix doesn't work.

MasterLongevities of 3h and 4h hit this on regular basis last one is 666.development-0.20191029.76581e7f147

```
2019-10-30 04:07:17.965: (DisruptionEvent Severity.ERROR): type=end name=TruncateMonkey node=Node longevity-100gb-4h-master-db-node-9d0c5d76-2 [54.165.53.175 | 172.30.0.188] (seed: False) duration=41 error=Encountered a bad command exit code!
Command: 'cqlsh --no-color   --request-timeout=120 --connect-timeout=5  -e "TRUNCATE ks_truncate.standard1" 172.30.0.188 9042'
Exit code: 2
Stdout:
Stderr:
<stdin>:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)

```

**Comment by glommer** [comment]
note that my fix is only relevant if auto_snapshots is set to false.

If it is set to true (the default), then truncate implies flush, which can
take a long while.

**Comment by amoskong** [comment]
Reproduced this issue in upgrade from 3.0.10-0.20190815.b3bfd8c08 to 3.1.2-0.20191127.9d78d848e
- job[ link](https://jenkins.scylladb.com/job/scylla-staging/job/amos/job/rolling-upgrade-clone-3.1/job/rolling-upgrade-debian9/4/execution/node/35/log/)
- the error raised in verify testing after upgraded one node.

**Comment by slivne** [comment]
the patch 889f31f is merged - lets see if we still hit this in 3.1 / 3.2 testing if not we can close

**Comment by yarongilor** [comment]
Reproduced in ICS 1TB longevity
(https://jenkins.scylladb.com/view/enterprise-2019.1/job/enterprise-2019.1/job/SCT_Enterprise_Feature/job/ICS/job/ics-longevity-1tb-7days-test/1/execution/node/43/log/?consoleFull)
```
Scylla version (or git commit hash): 2019.1.3-0.20191217.cb307772b
Cluster size: 4
OS (RHEL/CentOS/Ubuntu/AWS AMI): eu-west-1 - ami-0b8fd266fc7868428
```
```
(DisruptionEvent Severity.ERROR): type=end name=NodetoolCleanupMonkey node=Node longevity-tls-1tb-7d-2019-1-db-node-478a4bf8-2 [34.255.6.20 | 10.0.231.207] (seed: False) duration=28430 error=Encountered a bad command exit code!

Command: 'nodetool -u cassandra -pw cassandra cleanup keyspace_lz4'

Exit code: -1

Stdout:



Stderr:

WARN 04:58:31,136 Small cdc volume detected at /cdc_raw; setting cdc_total_space_in_mb to 1918. You can override this in cassandra.yaml


Traceback (most recent call last):
File "/sct/sdcm/nemesis.py", line 1117, in wrapper
result = method(*args, **kwargs)
File "/sct/sdcm/nemesis.py", line 1332, in disrupt
self.call_random_disrupt_method()
File "/sct/sdcm/nemesis.py", line 624, in call_random_disrupt_method
disrupt_method()
File "/sct/sdcm/nemesis.py", line 648, in disrupt_nodetool_cleanup
node.run_nodetool(sub_cmd="cleanup", args=keyspace)
File "/sct/sdcm/cluster.py", line 2198, in run_nodetool
result = self.remoter.run(cmd, ignore_status=ignore_status, verbose=verbose)
File "/sct/sdcm/remote.py", line 254, in run
result = _run()
File "/sct/sdcm/utils/common.py", line 88, in inner
return func(*args, **kwargs)
File "/sct/sdcm/remote.py", line 241, in _run
watchers=watchers, timeout=timeout)
File "", line 2, in run
File "/usr/lib/python2.7/site-packages/fabric/connection.py", line 30, in opens
return method(self, *args, **kwargs)
File "/usr/lib/python2.7/site-packages/fabric/connection.py", line 721, in run
return self._run(self._remote_runner(), command, **kwargs)
File "/usr/lib/python2.7/site-packages/invoke/context.py", line 101, in _run
return runner.run(command, **kwargs)
```

**Comment by ShlomiBalalis** [comment]
### Reproduced in 50gb-4days:
Scylla version (or git commit hash):3.2.rc3-0.20191215.39b17be5620
Cluster size:6
OS (RHEL/CentOS/Ubuntu/AWS AMI):ami-0cab4c6eceb3249ac

https://jenkins.scylladb.com/view/scylla-3.2/job/scylla-3.2/job/longevity/job/longevity-50gb-4days/5/console

```
DisruptionEvent Severity.ERROR): type=end name=TruncateMonkey node=Node longevity-tls-50gb-4d-3-2-db-node-6a8a4315-2 [52.30.50.100 | 10.0.215.34] (seed: False) duration=311 error=Encountered a bad command exit code!

Command: 'cqlsh --no-color -u cassandra -p cassandra  --request-timeout=120 --connect-timeout=5  -e "TRUNCATE ks_truncate.standard1" 10.0.215.34 9042'

Exit code: 2

Stdout:



Stderr:

:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)


Traceback (most recent call last):
  File "/sct/sdcm/nemesis.py", line 1181, in wrapper
    result = method(*args, **kwargs)
  File "/sct/sdcm/nemesis.py", line 1396, in disrupt
    self.call_random_disrupt_method()
  File "/sct/sdcm/nemesis.py", line 622, in call_random_disrupt_method
    disrupt_method()
  File "/sct/sdcm/nemesis.py", line 669, in disrupt_truncate
    self.target_node.run_cqlsh(cmd='TRUNCATE {}.{}'.format(keyspace_truncate, table), timeout=120)
  File "/sct/sdcm/cluster.py", line 2323, in run_cqlsh
    verbose=verbose)
  File "/sct/sdcm/remote.py", line 254, in run
    result = _run()
  File "/sct/sdcm/utils/common.py", line 90, in inner
    return func(*args, **kwargs)
  File "/sct/sdcm/remote.py", line 241, in _run
    watchers=watchers, timeout=timeout)
  File "", line 2, in run
  File "/usr/lib/python2.7/site-packages/fabric/connection.py", line 30, in opens
    return method(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/fabric/connection.py", line 721, in run
    return self._run(self._remote_runner(), command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/context.py", line 101, in _run
    return runner.run(command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/runners.py", line 291, in run
    return self._run_body(command, **kwargs)
  File "/usr/lib/python2.7/site-packages/invoke/runners.py", line 442, in _run_body
    raise UnexpectedExit(result)
UnexpectedExit: Encountered a bad command exit code!

Command: 'cqlsh --no-color -u cassandra -p cassandra  --request-timeout=120 --connect-timeout=5  -e "TRUNCATE ks_truncate.standard1" 10.0.215.34 9042'

Exit code: 2

Stdout:



Stderr:

:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)
```

**Comment by bentsi** [comment]
reproduced again on `3.2.0-0.20200113.798357f656`

Scylla YAML
```
# truncate_request_timeout_in_ms: 60000
# The default timeout for other, miscellaneous operations
# request_timeout_in_ms: 10000
```
Truncate command
```
< t:2020-01-13 19:53:51,702 p:INFO  > [centos@172.30.0.133]: Command "nodetool  flush " finished with status 0
< t:2020-01-13 19:53:51,702 p:DEBUG > Node longevity-100gb-4h-3-2-db-node-a94b42c6-6 [3.87.55.244 | 172.30.0.133] (seed: False): Command 'nodetool  flush ' duration -> 9.95304608345 s
< t:2020-01-13 19:53:51,703 p:DEBUG > [centos@172.30.0.133]: Running command "cqlsh --no-color   --request-timeout=120 --connect-timeout=5  -e "TRUNCATE ks_truncate.standard1" 172.30.0.133 9042"...

< t:2020-01-13 19:54:03,907 p:ERROR > [centos@172.30.0.133]: Error executing command: "cqlsh --no-color   --request-timeout=120 --connect-timeout=5  -e "TRUNCATE ks_truncate.standard1" 172.30.0.133 9042"; Exit status: 2
< t:2020-01-13 19:54:03,908 p:DEBUG > [centos@172.30.0.133]: STDERR: <stdin>:1:ProtocolException: Error during truncate: seastar::rpc::timeout_error (rpc call timed out)
```
DB log during the command:
```
2020-01-13T19:53:48+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 12] compaction - Compacted 2 sstables to [/var/lib/scylla/data/system_schema/aggregates-924c55872e3a345bb10c12f37c1ba895/mc-54-big-Data.db:level=0, ]. 10752 bytes to 5688 (~52% of original) in 42ms = 0.13MB/s. ~256 total partitions merged to 1.
2020-01-13T19:53:51+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 2] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27036-big-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26854-big-Data.db:level=0, ]
2020-01-13T19:53:51+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 11] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26849-big-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26653-big-Data.db:level=0, ]
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacting [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-87-big-Data.db:level=0, /var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-73-big-Data.db:level=0, ]
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacting [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-115-big-Data.db:level=0, /var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-101-big-Data.db:level=0, ]
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacted 2 sstables to [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-143-big-Data.db:level=0, ]. 11024 bytes to 5773 (~52% of original) in 20ms = 0.28MB/s. ~256 total partitions merged to 1.
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacting [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-129-big-Data.db:level=0, /var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-143-big-Data.db:level=0, ]
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacted 2 sstables to [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-157-big-Data.db:level=0, ]. 10629 bytes to 5717 (~53% of original) in 24ms = 0.23MB/s. ~256 total partitions merged to 1.
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacted 2 sstables to [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-171-big-Data.db:level=0, ]. 11008 bytes to 5775 (~52% of original) in 20ms = 0.28MB/s. ~256 total partitions merged to 1.
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacting [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-157-big-Data.db:level=0, /var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-171-big-Data.db:level=0, ]
2020-01-13T19:53:53+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 3] compaction - Compacted 2 sstables to [/var/lib/scylla/data/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-185-big-Data.db:level=0, ]. 11492 bytes to 5790 (~50% of original) in 18ms = 0.31MB/s. ~256 total partitions merged to 1.
2020-01-13T19:53:54+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 2] compaction - Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27050-big-Data.db:level=0, ]. 491037366 bytes to 491464336 (~100% of original) in 3438ms = 136.33MB/s. ~93696 total partitions merged to 93561.
2020-01-13T19:53:54+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 2] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26840-big-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27050-big-Data.db:level=0, ]
2020-01-13T19:53:54+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 11] compaction - Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26863-big-Data.db:level=0, ]. 551597360 bytes to 552092357 (~100% of original) in 3780ms = 139.29MB/s. ~105344 total partitions merged to 105104.
2020-01-13T19:53:54+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 11] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26863-big-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26667-big-Data.db:level=0, ]
2020-01-13T19:54:00+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 2] compaction - Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27064-big-Data.db:level=0, ]. 898246418 bytes to 895013635 (~99% of original) in 5779ms = 147.70MB/s. ~171008 total partitions merged to 170391.
2020-01-13T19:54:00+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 2] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27064-big-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27022-big-Data.db:level=0, ]
2020-01-13T19:54:02+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 11] compaction - Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26877-big-Data.db:level=0, ]. 1191579029 bytes to 1187900933 (~99% of original) in 7425ms = 152.58MB/s. ~227072 total partitions merged to 226153.
2020-01-13T19:54:02+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 11] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26835-big-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-26877-big-Data.db:level=0, ]
2020-01-13T19:54:12+00:00  ip-172-30-0-133 !INFO    | scylla: [shard 2] compaction - Compacted 2 sstables to [/var/lib/scylla/data/keyspace1/standard1-aa085610362e11ea81d1000000000000/mc-27078-big-Data.db:level=0, ]. 1861388139 bytes to 1591998467 (~85% of original) in 11620ms = 130.66MB/s. ~354944 total partitions merged to 303076.
```

It looks like the timeout for the RPC command should be 60 seconds but the command failed only after about ~12 seconds. That could probably mean that Scylla incorrectly handled timeout.
CC: @elcallio

**Comment by slivne** [comment]
@elcallio can you please check what bentsi has wrote .... (e.g. timeout value is not adhered)

**Comment by slivne** [comment]
@gleb-cloudius can you please help here - how is the timeout of truncate being 60 seconds translated to rpc timeout ?

**Comment by glommer** [comment]
12 seconds is the client timeout. If it failed in exactly 12 seconds, that means very likely the client timed this out.

**Comment by slivne** [comment]
@gleb-cloudius - help ... we see an rpc-timeout, so while glauber may be correct I can't connect the dots to the error message :)

**Comment by gleb-cloudius** [comment]
I failed to understand what this bug is about. The initial report has
this unfounded expectation that cqlsh's request-timeout parameter has
something to do with timeouting requests inside a server. It does not.
It configures client's timeout. Truncate timeout is configured by
truncate_request_timeout_in_ms in scylla.yaml and it defaults to 10
seconds.

So as far as I am concerned there are two issue here. One is that
internal rpc timeout is propagated to a client instead of been handled
internally and converted to proper cql error. Another is that truncate
is slow. Neither of them has nothing to do with RPC per se.


--
			Gleb.

**Comment by slivne** [comment]
@elcallio
-  can you try and reduce the truncate timeout (very low) see that you get the rpc exception and then convert it to the correct error.


with regards to being slow - truncate by default takes a snapshot in that case it needs to flush memtables and this can take time.

**Comment by roydahan** [comment]
FWIW, We added a nodetool flush just before the truncate and it didn't help.

**Comment by elcallio** [comment]
@slivne, I don't get what you want me to do here. Repro an rpc timeout in truncate? Correct error? What is a "correct" error. Afaict, the error returned is exactly what happened, no? RPC timed out. @gleb-cloudius says "converted to proper cql error" - does that mean just wrap it in cql timeout or what?

Flush is not the only thing potentially taking time for truncation. It needs to lock out compaction, it has to delete files on disk and rebuild stats  (though should be none left, so hopefully not expensive).

**Comment by gleb-cloudius** [comment]
We can check what Cassandra returns here and do the same. I am certain they do not
return rpc error.

--
			Gleb.

**Comment by elcallio** [comment]
We do _almost_ exactly what cassandra does: send an RPC (slightly different, but same principle), catch any exceptions, wrap them in `truncation_exception`. Difference is that for some reason our truncation exception sets the error code `PROTOCOL_ERROR` instead of `TRUNCATE_ERROR` - which would seem more appropriate.

**Comment by slivne** [comment]
we also fixed d7a9fc3611a96b862d35166763897dcf5708c848

I am closing this for now
